### ISSUE TITLE
Irina: State, 2 weeks

### ISSUE THREAD

**Issue by IrinaSing** [HackYourFutureBelgium/class-13-14#296]
> [module repository](_)

## Study Plan

<details>
<summary>expand/collapse</summary>
<br>

### Coding

- [What is state and why should we care about it?](https://dev.to/codeartistryio/what-is-state-and-why-should-we-care-about-it-4o95)
- Code in this Repo
  - 🥚 [function-roles](./function-roles): Learn about **components**, functions that _render_ data into DOM elements
  - 🥚 [stepped](./stepped) (examples): single-file projects that have been built one step at a time
  - 🥚 [refactors](./refactors) (examples): study refactored code from a single JS file to separated JS projects
  - 🐣 [separated](./separated) (examples): projects to study that are separated into many folders and files.
  - 🐣 [rendering-data](./rendering-data) (exercises): write functions that turn JS data into DOM elements, practice inspecting elements in the DOM and debugger to check your work.
  - 🐥 [testing-components](./testing-components) (exercises): learn how to read and pass tests for component functions(tests in this directory must be run in Node.js, they will not work in the browser)
- More Practice
  - [javascript-toy\_\_dom-exercises](https://github.com/omahacodeschool/javascript-toy__dom-exercises)
  - [practice-and-learn-the-dom](https://github.com/tutsplus/practice-javascript-and-learn-the-dom)
  - [Learn DOM Rendering](https://github.com/oliverjam/learn-dom-rendering)
  - [Learn Form Validation](https://github.com/oliverjam/learn-form-validation)
- Working with the DOM
  - [Traversy DOM Crash Course](https://www.youtube.com/watch?v=0ik6X4DJKCc) - it's long, no need to finish it before Sunday
  - [JS Dom Tutorials: Accessing The DOM](https://dom-tutorials.appspot.com/static/index.html)
  - [DOM Tree (js.info)](https://javascript.info/dom-nodes)
  - [Walking the DOM (js.info)](https://javascript.info/dom-navigation)
  - [getElement & querySelector (js.info)](https://javascript.info/searching-elements-dom)
- DOM Elements: Attributes vs. Properties
  - [javascript.info](https://javascript.info/dom-attributes-and-properties)
  - [FullstackTraning](https://blog.fullstacktraining.com/html-attributes-vs-dom-properties/) (article)
  - [WebTunings](https://www.youtube.com/watch?v=9DWKjn969lM) (video)
  - StackOverflow: [the difference](https://stackoverflow.com/questions/6003819/what-is-the-difference-between-properties-and-attributes-in-html), [when to use which](https://stackoverflow.com/questions/3919291/when-to-use-setattribute-vs-attribute-in-javascript)
- Nodes vs. Elements
  - [Web Tunings](https://www.youtube.com/watch?v=FrX1bzWc8O4)
  - [StackOverflow](https://stackoverflow.com/questions/9979172/difference-between-node-object-and-element-object)
- [What is Logging?](https://www.bmc.com/blogs/monitoring-logging-tracing/)

### Project Scoping

- What is project scope, and how to plan it
  - [be SMART](https://www.fool.com/the-blueprint/project-scope/)
  - [6 steps](https://www.potential.com/articles/project-scope/)
  - [cio.com](https://www.cio.com/article/3542776/what-is-project-scope-defining-and-outlining-project-success.html)
  - [clickup.com](https://clickup.com/blog/project-scope/)
- [What about Project Scope Management?](https://www.youtube.com/watch?v=cXAL56knUpE)
- [Requirements Gathering](https://www.brighthubpm.com/agile/123103-assembling-a-toolbox-for-agile-requirements/)
- [Scope Creep](https://duckduckgo.com/?q=scope+creep&atb=v214-1&ia=web)
- [Scope Statement Statement](https://duckduckgo.com/?q=project+scope+statement&atb=v214-1&ia=web)

</details>
<br>

Week 1

- [ ] started the exercises
- [ ] [project issue](your-link-here) (if there is a project)

**Comment by colevandersWands** [comment]
> No clear idea on how to implement data in real projects.

You'll learn much more about data in web pages when you reach Asynchronous Programming and start using data from APIs.  In web apps data can get pretty complicated pretty quickly because it's often stored in many different servers and needs to be used on many devices at the same time.  

For now, it's enough to get the idea that state is the data at each moment in your program, that state is always changing, and that you should plan your web page around the data it uses.

> Group project did not work this time.

What's not working about the project?

> I would like to know more about logic of writing data (why to choose these parameters to be data, not others).
Are there more external recourses on this topic?

The starting point for planning your data is to understand the domain you're working in.  Data should be somehow an abstract representation of what you're trying to achieve so there is no one-size-fits-all answer for how to choose. Beyond that some of the choices will depend on how you want to use the data and your technical constraints. 

And that's about all I can say on that for now without making things up.  I'll keep an eye out for good resources, but you'll also get more chances to explore this in the coming modules

**Comment by IrinaSing** [comment]
> What's not working about the project?

Lack of knowledge of what to do was most important in my opinion. Not much commitment, coordination and enthusiasm as well.  At this point of time it was more effective to study alone on the similar project.

**Comment by IrinaSing** [comment]
> And that's about all I can say on that for now without making things up. I'll keep an eye out for good resources, but you'll also get more chances to explore this in the coming modules

Thank you. I try to study examples but still can not extract patterns of workflow from them that I could use.

**Comment by colevandersWands** [comment]
> Lack of knowledge of what to do was most important in my opinion.

It's too soon for a group project like this, an individual version first would be better?  Or is it more a lack of material and guidance?

**Comment by IrinaSing** [comment]
> It's too soon for a group project like this, an individual version first would be better? Or is it more a lack of material and guidance?

In my opinion, individual project before group project is a good idea. Making to-do list with Traversy tutorial before starting the group project can be helpful to have a common ground. It is the first time when we have to write a program and it's difficult to plan, because we have to imagine already what should be done. Planning is still perceived as something unnecessary, by the way (thing that only takes time). A workshop on soft skills on principles of team work, distributing tasks etc. would also be helpful (as far as we are trying to prepare for real work). 

I think that exercises on DOM suits the previous module more that the current module.

**Comment by colevandersWands** [comment]
> I think that exercises on DOM suits the previous module more that the current module.

you mean the component functions?

**Comment by IrinaSing** [comment]
No, I mean resources from [Suggested study](https://github.com/HackYourFutureBelgium/state#coding)
And video is also from that list:

> Working with the DOM
> Traversy DOM Crash Course - it's long, no need to finish it before Sunday

**Comment by IrinaSing** [comment]
### Check-In

#### What went well?
~~~~
#### What went less well?
I did not finish my to-do list and I am very confused by complex interactions of code pieces in a program. 

#### I Need Help With:
Do I understand correct that the handler takes the data and insert it into component? Or can component to do actions with data? 
#### Lessons Learned

### Sunday Prep Work

**Comment by colevandersWands** [comment]
> I am confused by complex interactions of code pieces in a program.

Nothing to do about this but lots of practice.  Programs will only get more complex, and often each app will have a totally different structure than the last you worked on.  No (reasonable) person will expect you to understand a new project in a few days, so being confident you can figure it out with enough time is what counts.
